rust/integration-testing/src/bin/flight-test-integration-client.rs

```diff
@@ -0,0 +1,377 @@
+
+            if value != "expected value" {
+                let msg = format!(
+                    "Expected to receive header 'x-middleware: expected value', \
+                    but instead got: '{}'",
+                    value
+                );
+                return Err(Box::new(Status::internal(msg)));
+            }
+
+            eprintln!("Headers received successfully on failing call.");
+        }
+    }
+
+    // This call should succeed
+    descriptor.cmd = b"success".to_vec();
+    let resp = client.get_flight_info(Request::new(descriptor)).await?;
+
+    let headers = resp.metadata();
+    let middleware_header = headers.get("x-middleware");
+    let value = middleware_header.map(|v| v.to_str().unwrap()).unwrap_or("");
+
+    if value != "expected value" {
+        let msg = format!(
+            "Expected to receive header 'x-middleware: expected value', \
+            but instead got: '{}'",
+            value
+        );
+        return Err(Box::new(Status::internal(msg)));
+    }
+
+    eprintln!("Headers received successfully on passing call.");
+
+    Ok(())
+}
+
+fn middleware_interceptor(mut req: Request<()>) -> Result<Request<()>, Status> {
+    let metadata = req.metadata_mut();
+    metadata.insert("x-middleware", "expected value".parse().unwrap());
+    Ok(req)
+}
+
+async fn auth_basic_proto_scenario(host: &str, port: &str) -> Result {
+    let url = format!("http://{}:{}", host, port);
+    let mut client = FlightServiceClient::connect(url).await?;
+
+    let action = arrow_flight::Action::default();
+
+    let resp = client.do_action(Request::new(action.clone())).await;
+    // This client is unauthenticated and should fail.
+    match resp {
+        Err(e) => {
+            if e.code() != tonic::Code::Unauthenticated {
+                return Err(Box::new(Status::internal(format!(
+                    "Expected UNAUTHENTICATED but got {:?}",
+                    e
+                ))));
+            }
+        }
+        Ok(other) => {
+            return Err(Box::new(Status::internal(format!(
+                "Expected UNAUTHENTICATED but got {:?}",
+                other
+            ))));
+        }
+    }
+
+    let token = authenticate(&mut client, AUTH_USERNAME, AUTH_PASSWORD)
+        .await
+        .expect("must respond successfully from handshake");
+
+    let mut request = Request::new(action);
+    let metadata = request.metadata_mut();
+    metadata.insert_bin(
+        "auth-token-bin",
+        MetadataValue::from_bytes(token.as_bytes()),
+    );
+
+    let resp = client.do_action(request).await?;
+    let mut resp = resp.into_inner();
+
+    let r = resp
+        .next()
+        .await
+        .expect("No response received")
+        .expect("Invalid response received");
+
+    let body = String::from_utf8(r.body).unwrap();
+    assert_eq!(body, AUTH_USERNAME);
+
+    Ok(())
+}
+
+// TODO: should this be extended, abstracted, and moved out of test code and into production code?
```

Comment:
my intuition says no, as I'm assuming that there'd be different ways of implementing authentication.
I have however not used Flight, so if the basic auth scenario is always this, we could move this into `arrow-flight`.

Reply:
The C++ Flight client provides [this `Authenticate` function](https://github.com/apache/arrow/blob/cf243d4bfc496dedb21a9c155eaba10664266923/cpp/src/arrow/flight/client.cc#L977-L994) and the ability to specify a [`ClientAuthHandler`](https://github.com/apache/arrow/blob/dbeab70863c4d0cd3da800f18f7e474da624bb5f/cpp/src/arrow/flight/client_auth.h#L45-L59), but I'm not sure if that's the abstraction desired in Rust 🤷‍♀️

Reply:
I think we can always add it later to the core Rust library -- for now, having code in the example that does the work that people can use / follow seems pretty good to me